### ISSUE TITLE
Unable to read InputStream (AES-encrypted, CompressionMethod.STORE)

### ISSUE THREAD

**Issue by fkirc** [srikanth-lingala/zip4j#11]
Currently, Zip4J 2.X does not work at all for my AES encrypted files.
What I am trying to do is simple: Read a ZipInputStream from an AES-encrypted file entry like this:
```Java
final ZipInputStream is = zipFile.getInputStream(fileHeader);
do {
 n = is.read(buf, 0, buf_len);
 ... do some stuff with the buffer
} while (n != -1);
```
Note that I use `CompressionMethod.STORE`.
Depending on the file I am testing, the above snippet fails at one of multiple places.
My main test file has been created by Zip4j version 1.3.2.

- Firstly, the following line in `ZipInputStream:read()` looks suspicious:
```Java
int readLen = decompressedInputStream.read(b, off, (len - len %16));
```
If len is smaller than 16, then the above line may lead to an infinite loop since `len - len % 16` evaluates to zero and nothing is read at all.

- Still staying within `ZipInputStream:read()`: In some cases, the following piece of code produced a nullpointer exception:
```Java
// localFileHeader may be null!!!
if (isZipEntryDirectory(localFileHeader.getFileName())) {
      return -1;
}
```

- If we manage to get until `DecompressedInputStream:read()`, then the following failure may occur:
```Java
@Override
public int read(byte[] b, int off, int len) throws IOException {
    if (compressedSize != -1) {
      if (bytesRead >= compressedSize) {
        return -1; // This may return as a failure because compressedSize is zero and bytesRead is also set to zero when trying to read the first time!!!
      }
      .....
```

- Furthermore, I worry about the way that a "LocalFileHeader" is retrieved from a "FileHeader".
In particular, the following method is called within `zipFile.getInputStream(fileHeader);`:

```Java
public static ZipInputStream createZipInputStream(ZipModel zipModel, FileHeader fileHeader, char[] password)
      throws ZipException {
    try {
      SplitInputStream splitInputStream = new SplitInputStream(zipModel.getZipFile(), zipModel.isSplitArchive(),
          zipModel.getEndOfCentralDirectoryRecord().getNumberOfThisDisk());
      splitInputStream.prepareExtractionForFileHeader(fileHeader);

      ZipInputStream zipInputStream = new ZipInputStream(splitInputStream, password);
      if (zipInputStream.getNextEntry() == null) {
        throw new ZipException("Could not locate local file header for corresponding file header");
      }
      return zipInputStream;
    } catch (IOException e) {
      throw new ZipException(e);
    }
  }
```
All FileHeader entries seem to be **ignored** except of the offset to the LocalFileHeader, which is then parsed. Note that many fields in the Zip format are redundant (fields being present in both the central directory file header and the corresponding local file header).
A debugging session with an old sample file revealed the following:
```
fileHeader.compressedSize = 31
fileHeader.uncompressedSize = 3
localFileHeader.compressedSize = 0
localFileHeader.uncompressedSize = 0
```
**localFileHeader should not have zero values for compressedSize and uncompressedSize**.
In fact, I would expect that these fields are equal to the "central directory fileHeader".
Instead, these zero values seem to mess up the decryption and input streaming code in various places.

I am sure that the above file headers correspond to the same file since most of the other fields are equal (timestamps, file name and other stuff).

**Comment by fkirc** [comment]
In the meantime, before I manage to resolve everything:

Can you provide the source code for Zip4J version 1.3.3?
Perhaps on a separate branch?
I can only find 1.3.2.
1.3.2 is vulnerable to ZipSlip and potentially misses other bugfixes from 1.3.3.

**Comment by srikanth-lingala** [comment]
I will go through all of the points you mentioned in detail in a while. In the mean time, I have uploaded 1.3.3 source to a [different repo][1]. A lot has changed between 1.3.3 and 2.x to even consider putting it in a different branch.

[1]: https://github.com/srikanth-lingala/zip4j_1.3.3

**Comment by srikanth-lingala** [comment]
And the only difference between 1.3.2 and 1.3.3 is the ZipSlip fix.

**Comment by fkirc** [comment]
Thank you for uploading it.
This will also be a good reference for legacy applications who want to patch against ZipSlip with only minimal changes to their existing code base.

**Comment by fkirc** [comment]
I will now provide you with additional information to assist you with debugging the above-mentioned local file header inconsistency failure:

I  used the following file for my own debugging session:
[autosync_securezipnotes.zip](https://github.com/srikanth-lingala/zip4j/files/3356223/autosync_securezipnotes.zip)
The password for all these AES-encrypted file entries is: `aaaaaaaa`

What is really worrisome:
I created this file with Zip4j 1.3.2.
This file can be seamlessly opened with the "GNOME Archive Manager" (Ubuntu).
However, Zip4j 2.0.X is unable to extract this file.

The reason is the above-mentioned file header inconsistency.
Using the `Hexinator` hex editor, **I confirmed that the size fields of the local file header are zero, whereas the corresponding central directory file header has non-zero size fields** (this is at least true for the "Passwords__....." file entry at the beginning of this file).
The fields of interest are `compressedSize` and `uncompressedSize`

I wonder how you handled this kind of inconsistencies with previous versions with Zip4j?

In light of these results, I recommend to use Zip4j 1.3.3 until this issue is clarified and fixed with respect to backward compatibility.
Even if previous versions did something, we should retain binary backward compatibility at all cost.

**Comment by srikanth-lingala** [comment]
First of all, I highly appreciate your efforts and your detailed analysis. It is only with such collaboration that projects like these are possible. Thank you.

Coming to the points about the differences in compressed and uncompressed sizes between localfileheader and fileheader: there is one fundamental change between 1.3.3 and 2.x. To explain this, I have to first write about zip file format. 

An entry (file) in a zip file has the following format:

[LOCAL_FILE_HEADER][....DATA....][DATA_DESCRIPTOR]

And the above format repeats for the number of files present in the zip file, and at the end of the file, there are some additional headers containing information of the overall zip file (number of entries, and a lot of other info). (Note: the actual entry structure can get a lot more complicated with differing scenarios, the above entry structure is a very simplified version, just enough to make my point here).

DATA_DESCRIPTOR in the above structure is an optional header block. This header block contains the crc, compressed size and uncompressed size of the entry. This header block can be used if these sizes are not known upfront. Imagine working with streams, in this case, you do not know the size of compressed data upfront. You only know that information after you finish writing the data. But since we are working with streams, we cannot go back to the point where localfileheaders were written to update that data (compressed and uncompressed sizes). It is in this scenario, that DATA_DESCRIPTOR can be used. In such cases, crc, compressed size and uncompressed size are marked as 0 in localfilheader and the actual data is available in either DATA_DESCRIPTOR or the fileheader. This is totally as per the zip specification. And all zip tools should handle that scenario.

In Zip4j 2.x, when a zip file is created with the ZipFile api (and not the streams), internally Zip4j uses RandomAccessFile to add entries to the zip file. This way, Zip4j is able to go back in the file and update the local file header with crc and compressed size information (uncompressed size is already known upfront as this is a file and not stream and is also written when localfilheader is first written). On the contrary Zip4j 1.3.3 always used the DATA_DESCRIPTOR even if that information is available upfront and the underlying stream allows to go back and write/update data. This is a fundamental difference between 1.3.3 and 2.x. This change had to be done, because when this information (crc, and file sizes) are available and can be updated in the stream, it is always better to avoid DATA_DESCRIPTOR. 

So, when you see file sizes and crc as 0 in localfileheader, especially those created with v1.3.3, it is as expected and the actual values are written in DATA_DESCRIPTOR in this case. Zip4j handles such scenarios and considers the data from fileheader or from DATA_DESCRIPTOR if applicable.

**Comment by srikanth-lingala** [comment]
With the information from the above post, I now summarise the various points reported in your first post in this issue. Please let me know if I missed something.

* **Zip file created with v1.3.3 cannot be decrypted with Zip4j 2.x.**

    Thanks for uploading the sample zip file. This will help a lot. I will investigate this and let you know

* **len % 16 in ZipInputStream:read() does not behave as expected when len is less than 16**

    Firstly, length has to be split into a byte array size which has to be a multiple of 16 because AES encryption works that way. It expects a non-16 byte block only at the end of the file. It was for this purpose that length has to be multiple of 16. However, I agree that if length is less than 16, it will result in an indefinite loop. I will fix this.

* **In some cases ZipInputStream:read() results in NPE**

    Do you have an example for me? It would greatly help to investigate this.

* **DecompressedInputStream:read() may return -1 because compressed size is 0**

    As mentioned in the above post, it depends on where Zip4j is currently passing the compressed size from. I would presume that it is passing it from fileheader in case of ZipFile API and it tries to read localfileheader in case of using streams, but I need to have a look at this again. I will update it when I have more info.

* **All FileHeader entries seem to be ignored except of the offset to the LocalFileHeader** 

    Yes, fileheaders are ignored, because the file pointer is at the localfileheader and reading it made more sense, since we anyway have to read the data, to get to the "data" part of the entry. This way, we can also make sure that the entry being extracted is in fact the same entry that the extraction was called for. And hence [this check][1].

I will probably create new issues here on github from the ones that need fixes from the above points. This makes it easier to track them.

[1]: https://github.com/srikanth-lingala/zip4j/blob/master/src/main/java/net/lingala/zip4j/io/inputstream/ZipInputStream.java#L177

**Comment by fkirc** [comment]
Thank you for addressing all my issues.


> In some cases ZipInputStream:read() results in NPE

I will provide further information once I manage to reproduce this.

I understand that DATA_DESCRIPTOR should be avoided whenever it is possible.
However, Zip4j 2.0.X must be able to correctly deal with those files.
In particular, an API user of Zip4j cannot make any assumptions on whether some given entry uses DATA_DESCRIPTOR.

As of now, `zipFile.getInputStream(fileHeader)` does not work if the entry uses DATA_DESCRIPTOR.

**Comment by srikanth-lingala** [comment]
Created a new issue #12  for:

> Zip file created with v1.3.3 cannot be decrypted with Zip4j 2.x.

**Comment by srikanth-lingala** [comment]
Added an [update to the comment][1]

[1]: https://github.com/srikanth-lingala/zip4j/issues/11#issuecomment-508379712

**Comment by srikanth-lingala** [comment]
Create a new issue #13 for:

> len % 16 in ZipInputStream:read() does not behave as expected when len is less than 16

**Comment by srikanth-lingala** [comment]
Fixed in v2.1.0
